test/e2e/baseline_test.go

```diff
@@ -691,3 +697,142 @@ func SetWaypoint(t framework.TestContext, ns string, name string, waypoint strin
+					localDst := dst
+					localSrc := src
+					opt := echo.CallOptions{
+						Port:    echo.Port{Name: "tcp"},
+						Scheme:  scheme.TCP,
+						Count:   5,
+						Timeout: time.Second,
+						Check:   check.OK(),
+						To:      localDst,
+					}
+
+					query := buildL4Query(localSrc, localDst)
+					stc.Logf("prometheus query: %#v", query)
+					err := retry.Until(func() bool {
+						stc.Logf("sending call from %q to %q", deployName(localSrc), localDst.Config().Service)
+						localSrc.CallOrFail(stc, opt)
+						reqs, err := prom.QuerySum(localSrc.Config().Cluster, query)
+						if err != nil {
+							stc.Logf("could not query for traffic from %q to %q: %v", deployName(localSrc), localDst.Config().Service, err)
+							return false
+						}
+						if reqs == 0.0 {
+							stc.Logf("found zero-valued sum for traffic from %q to %q: %v", deployName(localSrc), localDst.Config().Service, err)
+							return false
+						}
+						return true
+					}, retry.Timeout(15*time.Second), retry.BackoffDelay(1*time.Second))
+					if err != nil {
+						PromDiff(t, prom, localSrc.Config().Cluster, query)
+						stc.Errorf("could not validate L4 telemetry for %q to %q: %v", deployName(localSrc), localDst.Config().Service, err)
+					}
+				})
+			}
+		}
+	})
+}
+
+func buildL4Query(src, dst echo.Instance) prometheus.Query {
+	query := prometheus.Query{}
+
+	srcns := src.NamespaceName()
+	destns := dst.NamespaceName()
+
+	labels := map[string]string{
+		"reporter":                       "destination",
+		"connection_security_policy":     "mutual_tls",
+		"response_flags":                 "-",
+		"request_protocol":               "tcp",
+		"destination_canonical_service":  dst.ServiceName(),
+		"destination_canonical_revision": dst.Config().Version,
+		"destination_service":            fmt.Sprintf("%s.%s.svc.cluster.local", dst.Config().Service, destns),
+		"destination_service_name":       fmt.Sprintf("%s.%s.svc.cluster.local", dst.Config().Service, destns),
+		"destination_service_namespace":  destns,
+		"destination_principal":          "-",
+		"destination_version":            dst.Config().Version,
+		"destination_workload":           deployName(dst),
+		"destination_workload_namespace": destns,
+		"destination_cluster":            "Kubernetes",
+		"source_canonical_service":       src.ServiceName(),
+		"source_canonical_revision":      src.Config().Version,
+		"source_principal":               "-",
+		"source_version":                 src.Config().Version,
+		"source_workload":                deployName(src),
+		"source_workload_namespace":      srcns,
+		"source_cluster":                 "Kubernetes",
+	}
+
+	query.Metric = "kmesh_tcp_connections_opened_total"
+	query.Labels = labels
+
+	return query
+}
+
+func deployName(inst echo.Instance) string {
+	return inst.ServiceName() + "-" + inst.Config().Version
+}
+
+func PromDiff(t test.Failer, prom prometheus.Instance, cluster cluster.Cluster, query prometheus.Query) {
```

Comment:
Why not just use the `PromDiff` function in `istio.io/istio/tests/integration/telemetry` ?